Collect statistics about binding allocations / local variable set

tenderlove opened the pull request:
This commit collects statistics about how many binding objects are
allocated as well as the number of local variables set on bindings.
Statistics are output along with other YJIT stats.  Here is an example
of the output:

```
***YJIT: Printing runtime counters from yjit.rb***
Number of Bindings Allocated: 195
Number of Local Variable Set: 0
opt_send_without_block exit reasons:
          ivar_get_method    7515891 (40.4%)
       se_cc_klass_differ    3081330 (16.6%)
       iseq_argc_mismatch    1564578 ( 8.4%)
     se_receiver_not_heap    1557663 ( 8.4%)
                 ic_empty    1407064 ( 7.6%)
         optimized_method     995823 ( 5.4%)
          iseq_not_simple     819413 ( 4.4%)
             alias_method     706972 ( 3.8%)
                  bmethod     685253 ( 3.7%)
      callsite_not_simple     225983 ( 1.2%)
                 kw_splat      25999 ( 0.1%)
          ivar_set_method        902 ( 0.0%)
       cfunc_toomany_args        394 ( 0.0%)
           refined_method         42 ( 0.0%)
    cfunc_ruby_array_varg         29 ( 0.0%)
              invalid_cme          4 ( 0.0%)
leave exit reasons:
    se_finish_frame    4067107 (100.0%)
       se_interrupt         24 ( 0.0%)
getinstancevariable exit reasons:
               undef     121177 (100.0%)
    idx_out_of_range          5 ( 0.0%)
opt_aref exit reasons:
    (all relevant counters are zero)
compiled_iseq_count:         3944
main_block_code_size:     1.1 MiB
side_block_code_size:     0.6 MiB
vm_insns_count:        1137268516
yjit_exec_insns_count:  414015644
ratio_in_yjit:              26.7%
avg_len_in_yjit:              7.5
total_exit_count:        55491789
most frequent exit op:
    opt_send_without_block:   18587628 (33.5%)
        opt_getinlinecache:   11075822 (20.0%)
                      send:    4949300 (8.9%)
                     leave:    4067131 (7.3%)
                   defined:    3975196 (7.2%)
       setinstancevariable:    3567315 (6.4%)
               invokesuper:    2982163 (5.4%)
        getblockparamproxy:    2168852 (3.9%)
                 opt_nil_p:    2104524 (3.8%)
                  opt_aref:    2013858 (3.6%)
```

Running RailsBench allocates 195 binding objects but doesn't set any
local variables.

maximecb commented on proc.c:
```diff
@@ -615,6 +619,10 @@ bind_local_variable_set(VALUE bindval, VALUE sym, VALUE val)
     }
 
+#if RUBY_DEBUG
+    rb_yjit_collect_local_variable_set();
```
Can you rename this to `rb_yjit_collect_binding_set()` ?

maximecb commented on yjit.rb:
```diff
@@ -60,6 +60,9 @@ def _print_stats
       $stderr.puts("***YJIT: Printing runtime counters from yjit.rb***")
 
+      $stderr.puts "Number of Bindings Allocated: %d\n" % counters[:binding_allocation_count]
+      $stderr.puts "Number of Local Variable Set: %d\n" % counters[:local_variable_set_count]
```
Also change this to "Number of locals modified through binding". Otherwise it's confusing, because `setlocal` is setting local variables.

tenderlove replied:
Do we care to distinguish between local variable values being overwritten versus new local variables?  Right now this just tracks any local variable set (not whether or not there was already an existing value)

maximecb replied:
I think for now it's ok to not distinguish. Anything that messes with the layout of the stack frame or changes local types could be problematic for us.

tenderlove replied:
Cool.  Text is changed! 😄

maximecb commented on yjit_iface.c:
```diff
@@ -613,6 +617,18 @@ rb_yjit_collect_vm_usage_insn(int insn)
+void
+rb_yjit_collect_local_variable_set(void)
+{
+    local_variable_set++;
```
Also rename `local_variable_set`

XrXr commented on yjit_iface.c:
```diff
@@ -581,6 +583,8 @@ get_stat_counters(rb_execution_context_t *ec, VALUE self)
             name_reader = name_end;
         }
+        rb_hash_aset(hash, ID2SYM(rb_intern("binding_allocation_count")), LL2NUM(binding_allocations));
+        rb_hash_aset(hash, ID2SYM(rb_intern("local_variable_set_count")), LL2NUM(local_variable_set));
```
Instead of adding two new globals, can you add the counters to yjit_iface.h? That way they get reset by `YJIT.reset_stats!`. You would increment them with `yjit_runtime_counters.binding_allocations++;`

XrXr commented on yjit_iface.c:
```diff
@@ -581,6 +581,8 @@ get_stat_counters(rb_execution_context_t *ec, VALUE self)
             name_reader = name_end;
         }
+        rb_hash_aset(hash, ID2SYM(rb_intern("binding_allocation_count")), LL2NUM(yjit_runtime_counters.binding_allocations));
+        rb_hash_aset(hash, ID2SYM(rb_intern("local_variable_set_count")), LL2NUM(yjit_runtime_counters.binding_set));
```
Don't need these anymore. Automatically covered by the loop above.

tenderlove replied:
Ooops!  Thanks!

maximecb commented on yjit.rb:
```diff
@@ -60,6 +60,9 @@ def _print_stats
 
       $stderr.puts("***YJIT: Printing runtime counters from yjit.rb***")
 
+      $stderr.puts "Number of Bindings Allocated: %d\n" % counters[:binding_allocation_count]
```
Nitpick: This Doesn't Need to be in Titlecase 😅

maximecb commented:
Looks pretty good. We just need those tests to pass 👍

tenderlove committed:
Fix case

maximecb commented:
@tenderlove can you update the dependencies so that the "Check Dependencies" don't fail?

tenderlove committed:
Collect statistics about binding allocations / local variable set

This commit collects statistics about how many binding objects are
allocated as well as the number of local variables set on bindings.
Statistics are output along with other YJIT stats.  Here is an example
of the output:

```
***YJIT: Printing runtime counters from yjit.rb***
Number of Bindings Allocated: 195
Number of locals modified through binding: 0
opt_send_without_block exit reasons:
          ivar_get_method    7515891 (40.4%)
       se_cc_klass_differ    3081330 (16.6%)
       iseq_argc_mismatch    1564578 ( 8.4%)
     se_receiver_not_heap    1557663 ( 8.4%)
                 ic_empty    1407064 ( 7.6%)
         optimized_method     995823 ( 5.4%)
          iseq_not_simple     819413 ( 4.4%)
             alias_method     706972 ( 3.8%)
                  bmethod     685253 ( 3.7%)
      callsite_not_simple     225983 ( 1.2%)
                 kw_splat      25999 ( 0.1%)
          ivar_set_method        902 ( 0.0%)
       cfunc_toomany_args        394 ( 0.0%)
           refined_method         42 ( 0.0%)
    cfunc_ruby_array_varg         29 ( 0.0%)
              invalid_cme          4 ( 0.0%)
leave exit reasons:
    se_finish_frame    4067107 (100.0%)
       se_interrupt         24 ( 0.0%)
getinstancevariable exit reasons:
               undef     121177 (100.0%)
    idx_out_of_range          5 ( 0.0%)
opt_aref exit reasons:
    (all relevant counters are zero)
compiled_iseq_count:         3944
main_block_code_size:     1.1 MiB
side_block_code_size:     0.6 MiB
vm_insns_count:        1137268516
yjit_exec_insns_count:  414015644
ratio_in_yjit:              26.7%
avg_len_in_yjit:              7.5
total_exit_count:        55491789
most frequent exit op:
    opt_send_without_block:   18587628 (33.5%)
        opt_getinlinecache:   11075822 (20.0%)
                      send:    4949300 (8.9%)
                     leave:    4067131 (7.3%)
                   defined:    3975196 (7.2%)
       setinstancevariable:    3567315 (6.4%)
               invokesuper:    2982163 (5.4%)
        getblockparamproxy:    2168852 (3.9%)
                 opt_nil_p:    2104524 (3.8%)
                  opt_aref:    2013858 (3.6%)
```

Running RailsBench allocates 195 binding objects but doesn't set any
local variables.

tenderlove committed:
update deps
